src/tools/crossgen2/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs

```diff
@@ -42,6 +42,31 @@ public sealed class DependencyAnalyzer<MarkStrategy, DependencyContextType> : De
         private bool _markingCompleted = false;
         private Random _stackPopRandomizer = null;
 
+        private class RandomInsertStack
+        {
+            List<DependencyNodeCore<DependencyContextType>> _nodes = new List<DependencyNodeCore<DependencyContextType>>();
+
+            public DependencyNodeCore<DependencyContextType> Pop()
+            {
+                DependencyNodeCore<DependencyContextType> node = _nodes[_nodes.Count - 1];
+                _nodes.RemoveAt(_nodes.Count - 1);
+                return node;
+            }
+
+            public int Count => _nodes.Count;
+
+            public void Push(DependencyNodeCore<DependencyContextType> node)
+            {
+                _nodes.Add(node);
+            }
+
+            public void InsertAtRandom(Random randomizer, DependencyNodeCore<DependencyContextType> node)
```

Comment:
You could put the `Random` object inside `RandomInsertStack` by having the constructor take an optional `Random`.  If a `Random` is passed in, you get the random insertion behavior. If not, it behaves like a regular stack.  That would let you just use Pop / Push and get rid of `InsertAtRandom` completely. You could then get rid of the `if (_stackPopRandomizer == null)` block in `AddToMarkStack` and simply call `Push`.